### DISCUSSION TITLE
Manipulating the title slot in timeline_entry

### DISCUSSION THREAD

**Discussion by nkonts** [zauberzeug/nicegui#3294] [Q&A] [answered]
### Question

Hi,

I am trying NiceGui for a new app I am building and trying to figure out how I can accomplish the style I am looking for. 

Basically, I use a timeline to show the last $n$ text elements. Those have a different width and my goal is to have all `timeline_entry` texts be 
- centered instead of e.g. left-aligned as the is the default property 
- their container to have the same shape

Plus, I want to get rid of "too much free space": When I add my items into the body of the entry and dont have a title, there is still some empty space between the text and subtitle where the title would be. I want to get rid of the space the title creates. An alternative would be to add my items as a title and manipulate them that way but then I cant get them centered inside containers with equal width.

Those are the best I could create so far. I simply use `ui.card` or `ui.button` to see the bounds of the container my `ui.label` is in:

```py
def signal_timeline(country: str, n: int) -> None:

    last_signals = {
        pd.Timestamp('2021-01-01'): "Something very long", 
        pd.Timestamp('2021-02-01'): "Something", 
        pd.Timestamp('2021-03-01'): "Short"
    }
    
    # The buttons (same with ui.card or ui.column) have unequal width
    with ui.timeline(side="right").classes("ml-4"):
        for i, (index, value) in enumerate(last_signals.items()):
            # Manipulate the entry title: https://github.com/zauberzeug/nicegui/discussions/2763
            entry = ui.timeline_entry(
                subtitle=index.strftime('%B, %Y'),
                icon="equalizer" if i == 0 else None  # icon for the first entry
            )
            with entry.add_slot('title'):
                ui.button(value).classes('w-full text-center')


    # Doesnt work but I dont have experience with html+css
    with ui.timeline(side="right").classes("ml-4"):
        for i, (index, value) in enumerate(last_signals.items()):
            # Manipulate the entry title: https://github.com/zauberzeug/nicegui/discussions/2763
            entry = ui.timeline_entry(
                subtitle=index.strftime('%B, %Y'),
                icon="equalizer" if i == 0 else None  # icon for the first entry
            )
            with entry.add_slot('title'):
                ui.html(f"<div class='w-full bg-blue-100 text-center font-bold'>{value}</div>")
            

    # This body looks like how I want it to but I cant get rid of the space the empty title is creating
    with ui.timeline(side="right").classes("ml-4"):
        for i, (index, value) in enumerate(last_signals.items()):
            # Manipulate the entry title: https://github.com/zauberzeug/nicegui/discussions/2763
            with ui.timeline_entry(
                subtitle=index.strftime('%B, %Y'),
                icon="equalizer" if i == 0 else None  # icon for the first entry
            ):
                with ui.card().classes("no-shadow border-[1px] w-full items-center"):
                    ui.label(value)
```

**Comment by falkoschindler** [comment]
Hi @nkonts,

The problem seems to be that Quasar's Timeline comes with 100 % width by default. By adding `.classes('w-auto')` to `ui.timeline` the timeline shrinks to a reasonable size while all title cards keep having equal width.

**Reply by nkonts** [reply]
That wasnt really what I wanted to achieve but its helpful since I didnt think to manipulate the classes of `ui.timeline` itself. To better illustrate my goal:

Center all the elements in the title slots and them same width:
<img width="239" alt="image" src="https://github.com/zauberzeug/nicegui/assets/32848650/97993668-5b72-43e6-99f8-93436d900b8d">
```py
    with ui.timeline(side="right").classes("ml-4 w-auto"):
        for i, (index, value) in enumerate(last_signals.items()):
            entry = ui.timeline_entry(
                subtitle=index.strftime('%B, %Y'),
                icon="equalizer" if i == 0 else None  # icon for the first entry
            )
            with entry.add_slot('title'):
                with ui.card().classes("no-shadow border-[1px] w-full items-center"):
                    ui.label(value)
```

or have an empty title but remove the space it creates market in red (here you can see that the width is the same for each body element of `timeline_entry`, exactly how I would like it to have):
<img width="184" alt="image" src="https://github.com/zauberzeug/nicegui/assets/32848650/ac14f1c3-5fcf-4c84-8246-c2be16a07439">

```py
    with ui.timeline(side="right").classes("ml-4 w-auto"):
        for i, (index, value) in enumerate(last_signals.items()):
            with ui.timeline_entry(
                subtitle=index.strftime('%B, %Y'),
                icon="equalizer" if i == 0 else None  # icon for the first entry
            ):
                with ui.card().classes("no-shadow border-[1px] w-full items-center"):
                    ui.label(value)
```

I tried `ui.timeline(side="right").classes("ml-4 w-auto iems-center text-center")` but this seems to manipulate the `subtitle` container and not `title`

**Reply by falkoschindler** [reply, marked as answer]
Ah, I see. Quasar's title slot doesn't fill its parent's width. A piece of custom CSS might be a quick workaround:
```py
ui.add_css('''
.q-timeline__title {
    width: 100%;
}
''')
```

Or you don't use the title slot and prevent it from taking up space:
```py
ui.add_css('''
.q-timeline__title {
    display: none;
}
''')
```

Or you add the class "-mt-8" to `ui.card` to add some negative top margin. This will move the card up to cover the unused title slot. This is a bit hacky, but should work nonetheless.

**Reply by nkonts** [reply]
Works perfectly, thanks! The hacky solutions are sometimes still very useful to learn understand the details of what is possible and what isnt 😄
